### PR TITLE
Svtdanny/t5 modeling

### PR DESCRIPTION
**Pull request by svtdanny**

### CODE REVIEW

File: notebooks/t5_sequence_modeling.ipynb [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+   "source": [
+    "!python3 ../wrappers/python/examples/causal_lm_data_preparation.py \\\n",
+    "    --hf-tokenizer=\"google/flan-t5-small\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Below we show an example of utilizing the GPT-2 model, implemented using the NNTile framework. We explore the following scenarios:\n",
```
Comment (Muxas) [inline notebooks/t5_sequence_modeling.ipynb:164]: GPT-2 -> T5

File: notebooks/t5_sequence_modeling.ipynb [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    "- **Training the model from a random initial state and saving it to a checkpoint.**\n",
+    "- **Loading the model weights from a checkpoint and continuing training with a different data type.**\n",
+    "- **Training the remote model downloaded from the Hugging Face infrastructure.**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For training and continuing retraining scenarios, the interpreted file \"gpt2_lmhead_training.py\" is used. This Python script supports the following arguments when running:\n",
```
Comment (Muxas) [inline notebooks/t5_sequence_modeling.ipynb:175]: `gpt2_lmhead_training.py` -> proper file

File: wrappers/python/nntile/layer/t5_attention.py [hunk 3]
```diff
@@ -0,0 +1,1580 @@
...
+        inner_dim = n_emb if inner_dim is None else inner_dim
+        inner_dim_tile = n_emb_tile if inner_dim_tile is None else inner_dim_tile
+
+        head_size = inner_dim // n_head
+        # Stupid check, that is not necessary, as the code shall work
+        if inner_dim != head_size * n_head:
+            raise RuntimeError
+        n_emb_k = x_k.value.shape[0]
+        n_emb_k_tile = x_k.value.basetile_shape[0]
+        # if [n_seq, n_batch] != x_k.value.shape[1:]:
```
Comment (Muxas) [inline wrappers/python/nntile/layer/t5_attention.py:308]: are these checks unnecessary?

File: wrappers/python/nntile/layer/t5_attention.py [hunk 4]
```diff
@@ -0,0 +1,1580 @@
...
+            max_distance=128,
+        )
+        relative_position_bucket_np_value = relative_position_bucket_np_value[None, ...]
+        if self.a.value.shape[2] > 1:
+            relative_position_bucket_np_value = (
+                relative_position_bucket_np_value.repeat(self.a.value.shape[2], axis=0)
+            )
+
+        self.relative_position_bucket_nnt = nntc.from_array(
+            relative_position_bucket_np_value.T,
```
Comment (Muxas) [inline wrappers/python/nntile/layer/t5_attention.py:966]: This tensor is of type int32, but embedding layer requires int64. There is no type Tensor_int32,  but input type np.int32 is mapped into Tensor_int64. There might some bugs, related to such a conversion.

File: wrappers/python/nntile/layer/t5_attention.py [hunk 5]
```diff
@@ -0,0 +1,1580 @@
...
+        # Q and K can be offloaded from GPU
+        self.q.value.wont_use()
+        self.k.value.wont_use()
+        # Calculate softmax inplace
+        # A = softmax(A, axis=0)
+        # Apply mask if needed
+        if self.mask:
+            mask_scalar_async(self.mask, self.val, self.a.value, 2)
+            # self.mask.wont_use()
+            # add_async(-1e30, self.mask, 1.0, self.a.value, self.a.value)
```
Comment (Muxas) [inline wrappers/python/nntile/layer/t5_attention.py:1022]: misleading commented operation

File: wrappers/python/nntile/model/t5_lmhead.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        activations.extend(act_fn_layer.activations_output)
+
+        # Output projection layer
+        out_proj, next_tag = Linear.generate_simple(
+            activations[-1],
+            "R",
+            notrans,
+            gemm_ndim,
+            [self.num_labels],
+            [self.num_labels],  # Using num_labels as tile size
```
Comment (Muxas) [inline wrappers/python/nntile/model/t5_lmhead.py:76]: Generally, we shall use special value for tiling of this linear layer. But since it is the same for other models, we shall stick with it for now until we completely rework Python API.

File: wrappers/python/nntile/model/t5_model.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        torch_model: T5ForSequenceClassificationTorch,
+        x: TensorMoments,
+        decoder_x: TensorMoments,
+        config: T5ConfigNNTile,
+        next_tag: int = 0,
+    ):
+        dtype2tensor_type = {
+            "fp32": Tensor_fp32,
+            "bf16": Tensor_bf16,
+            "fp32_fast_tf32": Tensor_fp32_fast_tf32,
```
Comment (Muxas) [inline wrappers/python/nntile/model/t5_model.py:160]: need to add types Tensor_fp32_fast_fp16 and Tensor_fp32_fast_bf16

File: wrappers/python/tests/layer/test_t5_attention.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    ],
+)
+@pytest.mark.parametrize(
+    "is_cross_attn",
+    [
+        True,
+        False,
+    ],
+)
+class TestT5Attention:
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_t5_attention.py:213]: test_coercion function is missing, that checks from_torch and to_torch conversions like [here](https://github.com/nntile/nntile/blob/5b8493f2a3a7bb3c2ee7359a35949afa9135533c/wrappers/python/tests/layer/test_gpt2_attention.py#L144)

File: wrappers/python/tests/layer/test_t5_attention.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+            ) <= rtol * torch.norm(torch_layer.relative_attention_bias.weight.grad)
+
+        if is_cross_attn:
+            encoder_output_grad = torch.Tensor(
+                to_numpy(nntile_layer.activations_input[1].grad).T
+            )
+            assert torch.norm(
+                encoder_output_grad - encoder_output_torch.grad
+            ) <= rtol * torch.norm(encoder_output_torch.grad)
+
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_t5_attention.py:281]: missing checking of entire attention module, only output is tested

File: wrappers/python/tests/model/test_t5_block.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+@pytest.mark.parametrize(
+    "dtype",
+    [
+        "fp32",
+        pytest.param("fp32_fast_tf32", marks=nocuda),
+        pytest.param("bf16", marks=nocuda),
+    ],
+)
+class TestT5Block:
+    def test_forward(
```
Comment (Muxas) [inline wrappers/python/tests/model/test_t5_block.py:215]: test_coercion is missing

File: wrappers/python/tests/model/test_t5_block.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        # NNTile forward and backward pass
+        nntile_block.forward_async()
+        nntile_block.backward_async()
+
+        # Compare gradients
+        grad_nntile = torch.Tensor(nntc.to_numpy(nntile_block.activations[0].grad).T)
+        rtol = dtype2tol[dtype]["rtol"]
+
+        assert torch.norm(x.grad - grad_nntile) <= rtol * torch.norm(x.grad)
+
```
Comment (Muxas) [inline wrappers/python/tests/model/test_t5_block.py:257]: need to check not only gradient of input, but also gradients of parameters

File: wrappers/python/tests/model/test_t5_lmhead.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+)
+@pytest.mark.parametrize(
+    "dtype",
+    [
+        "fp32",
+        pytest.param("fp32_fast_tf32", marks=nocuda),
+        pytest.param("bf16", marks=nocuda),
+    ],
+)
+class TestT5ClassificationHead:
```
Comment (Muxas) [inline wrappers/python/tests/model/test_t5_lmhead.py:175]: test_coercion

File: wrappers/python/tests/model/test_t5_lmhead.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        nntile_head.forward_async()
+        nntile_head.backward_async()
+
+        # Compare gradients for input
+        x_grad_nntile = torch.Tensor(nntc.to_numpy(nntile_head.activations[0].grad).T)
+        np.testing.assert_allclose(
+            x_torch.grad.detach().numpy(),
+            x_grad_nntile.detach().numpy(),
+            **dtype2tol[dtype],
+        )
```
Comment (Muxas) [inline wrappers/python/tests/model/test_t5_lmhead.py:215]: add testing of gradients of all parameters

File: wrappers/python/tests/model/test_t5_model.py [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+        )
+        enc_grad_nntile = torch.Tensor(
+            nntc.to_numpy(nntile_model.activations[0].grad).T
+        )
+
+        rtol = dtype2tol[dtype]["rtol"]
+
+        assert torch.norm(enc_x.grad - enc_grad_nntile) <= rtol * torch.norm(enc_x.grad)
+        assert torch.norm(dec_x.grad - dec_grad_nntile) <= rtol * torch.norm(dec_x.grad)
+
```
Comment (Muxas) [inline wrappers/python/tests/model/test_t5_model.py:357]: test gradients of all parameters

**Review by Muxas — COMMENTED** [review]
Testing to/from torch conversions and gradients of all parameters in backward are missing

**Review by Muxas — APPROVED** [review]
get green tick on automated tests and good to go
